src/app/codegen-data-model-provider/EmberAttributeDataBuffer.cpp

```diff
@@ -321,6 +375,200 @@ CHIP_ERROR EmberAttributeDataBuffer::Decode(chip::TLV::TLVReader & reader)
     return CHIP_NO_ERROR;
 }
 
+CHIP_ERROR EmberAttributeDataBuffer::EncodeInteger(chip::TLV::TLVWriter & writer, TLV::Tag tag, EndianReader & reader) const
+{
+    // Encodes an integer by first reading as raw bytes and then
+    // bitshift-convert
+    //
+    // This optimizes code size rather than readability at this point.
+
+    uint8_t raw_bytes[8];
+
+    bool isSigned = (mAttributeType == ZCL_INT8S_ATTRIBUTE_TYPE) //
+        || (mAttributeType == ZCL_INT16S_ATTRIBUTE_TYPE)         //
+        || (mAttributeType == ZCL_INT24S_ATTRIBUTE_TYPE)         //
+        || (mAttributeType == ZCL_INT32S_ATTRIBUTE_TYPE)         //
+        || (mAttributeType == ZCL_INT40S_ATTRIBUTE_TYPE)         //
+        || (mAttributeType == ZCL_INT48S_ATTRIBUTE_TYPE)         //
+        || (mAttributeType == ZCL_INT56S_ATTRIBUTE_TYPE)         //
+        || (mAttributeType == ZCL_INT64S_ATTRIBUTE_TYPE);
+
+    unsigned byteCount;
+    uint64_t nullValue;
+
+    if (isSigned)
+    {
+        const SignedDecodeInfo info = GetSignedDecodeInfo(mAttributeType);
+        byteCount                   = info.byteCount;
+        nullValue                   = static_cast<uint64_t>(info.minValue); // just a bit cast for easy compare
+    }
+    else
+    {
+        const UnsignedDecodeInfo info = GetUnsignedDecodeInfo(mAttributeType);
+        byteCount                     = info.byteCount;
+        nullValue                     = info.maxValue;
+    }
+
+    VerifyOrDie(sizeof(raw_bytes) >= byteCount);
+    if (!reader.ReadBytes(raw_bytes, byteCount).IsSuccess())
+    {
+        return reader.StatusCode();
+    }
+
+    // At this point, RAW_VALUE contains the actual value, need to make it "real"
+    union
+    {
+        int64_t int_value;
+        uint64_t uint_value;
+    } value;
+
+    value.uint_value = 0;
+
+#if CHIP_CONFIG_BIG_ENDIAN_TARGET
+    bool isNegative = isSigned && (raw_bytes[0] >= 0x80);
+    if (isNegative)
+    {
+        value.int_value = -1;
+    }
+    for (int i = 0; i < static_cast<int>(byteCount); i++)
+    {
+#else
+    bool isNegative = isSigned && (raw_bytes[byteCount - 1] >= 0x80);
+    if (isNegative)
+    {
+        value.int_value = -1;
+    }
+    for (int i = static_cast<int>(byteCount) - 1; i >= 0; i--)
+    {
+#endif
+        value.uint_value <<= 8;
+        value.uint_value = (value.uint_value & ~0xFFULL) | raw_bytes[i];
+    }
+
+    if (mIsNullable && (value.uint_value == nullValue))
+    {
+        // MaxValue is used for NULL setting
+        return writer.PutNull(tag);
+    }
+
+    switch (mAttributeType)
+    {
+    case ZCL_INT8U_ATTRIBUTE_TYPE: // Unsigned 8-bit integer
+        return writer.Put(tag, static_cast<uint8_t>(value.uint_value));
```

Comment:
Why are we doing this?  Assuming the value is in range at all, we can just `Put()` as a 64-bit signed/unsigned respectively depending on isSigned and don't need this whole big switch and the casting.